### PR TITLE
Validate runtime context entries in `runWith()` and `runWithSync()`

### LINKED ISSUE #424: `runWith()` and `runWithSync()` do not validate runtime context entries
## Summary

`runWith()` and `runWithSync()` assume every entry in `contexts` is a real `SourceContext`, but they never validate that runtime shape before dereferencing it.

## Reproduction

~~~~ typescript
import { runWith, runWithSync } from "@optique/core/facade";
import { option } from "@optique/core/primitives";
import { string } from "@optique/core/valueparser";

const parser = option("--name", string());

console.log(runWithSync(parser, "test", [{} as never], {
  args: [],
  help: false,
  version: false,
}));

console.log(await runWith(parser, "test", [{} as never], {
  args: [],
  help: false,
  version: false,
}));
~~~~

## Expected behavior

Malformed runtime context entries should be rejected with a clear validation error.

## Actual behavior

The malformed context is dereferenced directly and the runner crashes with raw errors such as `TypeError: context.getAnnotations is not a function`.

### PR DESCRIPTION
**Pull request by dahlia**
## Summary

Fixes https://github.com/dahlia/optique/issues/424.

`runWith()`, `runWithSync()`, and `runWithAsync()` previously assumed that every entry in the `contexts` array was a valid `SourceContext`, without any runtime validation. Passing a malformed object (e.g., `{}` or `null`) would cause the runner to crash with an unhelpful error like `TypeError: context.getAnnotations is not a function`.

This PR adds upfront validation so that invalid entries are rejected early with a clear `TypeError` that includes the actual type and index of the offending entry:

```typescript
runWithSync(parser, "test", [{} as never], { args: [] });
// TypeError: Expected each context to be a SourceContext, but got: object at index 0.
```

A new `isSourceContext()` predicate is also exported from *@optique/core/context* for use by external code that needs to check whether a value conforms to the `SourceContext` interface at runtime.

## Test plan

- [x] Added 7 tests covering `runWith` and `runWithSync` with various invalid context types: empty object, `null`, number, array, and correct index reporting for the second invalid entry in a multi-context array
- [x] All existing tests pass (`mise test:deno`, 360 tests)
- [x] Type check and lint pass (`mise check`)

### CODE REVIEW

File: packages/core/src/facade.test.ts [hunk 1]
```diff
@@ -9148,3 +9148,99 @@ describe("branch coverage: facade.ts edge cases", () => {
...
+    });
+
+    it("rejects with TypeError for array context", async () => {
+      await assert.rejects(
+        () => runWith(parser, "test", [[] as never], baseOptions),
+        {
+          name: "TypeError",
+          message:
+            /Expected each context to be a SourceContext, but got: array at index 0\./,
+        },
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts:9256]: Addressed in eadf0ff. Changed all RegExp assertions to exact string comparisons, including the second-invalid-context test which now asserts the full message.

File: packages/core/src/facade.test.ts [hunk 2]
```diff
@@ -9148,3 +9148,99 @@ describe("branch coverage: facade.ts edge cases", () => {
...
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts:9259]: Addressed in 9b720c4. Added a `runWithAsync` describe block with a null-context rejection test.
